src/corehost/cli/fx_reference.cpp

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "pal.h"
+#include "fx_ver.h"
+#include "fx_reference.h"
+#include "roll_fwd_on_no_candidate_fx_option.h"
+
+bool fx_reference_t::is_forward_compatible(const fx_ver_t& other) const
+{
+    // We expect the version to be <
+    assert(get_fx_version_number() < other);
+
+    // Support >= cases for release
+    if (get_fx_version_number() >= other)
+    {
+        return true;
+    }
+
+    if (get_use_exact_version())
+    {
+        return false;
+    }
+
+    // Verify major roll forward
+    if (get_fx_version_number().get_major() != other.get_major()
+        && roll_fwd_on_no_candidate_fx != roll_fwd_on_no_candidate_fx_option::major)
+    {
+        return false;
+    }
+
+    // Verify minor roll forward
+    if (get_fx_version_number().get_minor() != other.get_minor()
+        && roll_fwd_on_no_candidate_fx != roll_fwd_on_no_candidate_fx_option::major
+        && roll_fwd_on_no_candidate_fx != roll_fwd_on_no_candidate_fx_option::minor)
+    {
+        return false;
+    }
+
+    // Verify patch roll forward
+    // We do not distinguish here whether a previous framework reference found a patch version based on:
+    //  - initial reference matching a patch version,
+    //  - or roll_fwd_on_no_candidate_fx_option=major\minor finding a compatible patch version as initial framework,
+    //  - or applyPatches=true finding a newer patch version
+    if (get_fx_version_number().get_patch() != other.get_patch()
+        && patch_roll_fwd == false
+        && roll_fwd_on_no_candidate_fx != roll_fwd_on_no_candidate_fx_option::major
+        && roll_fwd_on_no_candidate_fx != roll_fwd_on_no_candidate_fx_option::minor)
+    {
+        return false;
+    }
+
+    // Release cannot roll forward to pre-release
+    if (!get_fx_version_number().is_prerelease() && other.is_prerelease())
+    {
+        return false;
+    }
+
+    return true;
+}
+
+void fx_reference_t::apply_settings_from(const fx_reference_t& from)
+{
+    if (from.get_fx_version().length() > 0)
+    {
+        set_fx_version(from.get_fx_version());
+    }
+
+    if (from.has_roll_fwd_on_no_candidate_fx)
+    {
+        set_roll_fwd_on_no_candidate_fx(*from.get_roll_fwd_on_no_candidate_fx());
+    }
+
+    if (from.has_patch_roll_fwd)
+    {
+        set_patch_roll_fwd(*from.get_patch_roll_fwd());
+    }
+}
+
+void fx_reference_t::merge_roll_forward_settings_from(const fx_reference_t& from)
+{
+    if (from.has_roll_fwd_on_no_candidate_fx)
+    {
+        if (!has_roll_fwd_on_no_candidate_fx ||
+            (*from.get_roll_fwd_on_no_candidate_fx()) < (*get_roll_fwd_on_no_candidate_fx()))
```

Comment:
I believe my thought pattern was:
- The public calls to set_ are there since we don't want to duplicate the logic in the setters.
- The public calls to get_ are there for consistency with the calls to set_
- The private calls to has_ are there for efficiency so we don't call get_ twice

However, I can remove the private calls to has_ and use a variable to store the value from get_ so we don't call it twice

Reply:
The reason for my comment was that I didn't understand the dereference at first - I expected the get_ to return a value directly, not a pointer. Then I realized that it effectively returns a nullable value and it's intentional. And then it took me a while to realize that the has_ are the underlying mechanism how we store the nullable values. Overall rather confusing.
I like the removal of has_ and using the public accessors only, leaving the has_ fields as really internal implementation details of the property itself and nothing else.